flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/JdbcExactlyOnceOptions.java

```diff
@@ -0,0 +1,130 @@
+ * number.
+ *
+ * <p><b>allowOutOfOrderCommits</b> - If true, all prepared transactions will be attempted to commit
+ * regardless of any transient failures during this operation. This may lead to inconsistency.
+ * Default: false.
+ *
+ * <p><b>recoveredAndRollback</b> - whether to rollback prepared transactions known to XA RM on
+ * startup (after committing <b>known</b> transactions, i.e. restored from state).
+ *
+ * <p>NOTE that setting this parameter to true may:
+ *
+ * <ol>
+ *   <li>interfere with other subtasks or applications (one subtask rolling back transactions
+ *       prepared by the other one (and known to it))
+ *   <li>block when using with some non-MVCC databases, if there are ended-not-prepared transactions
+ * </ol>
+ *
+ * See also {@link org.apache.flink.connector.jdbc.xa.XaFacade#recover()}
+ */
+public class JdbcExactlyOnceOptions implements Serializable {
+
+    private static final boolean DEFAULT_RECOVERED_AND_ROLLBACK = false;
+    private static final int DEFAULT_MAX_COMMIT_ATTEMPTS = 3;
+    private static final boolean DEFAULT_ALLOW_OUT_OF_ORDER_COMMITS = false;
+
+    private final boolean discoverAndRollbackOnRecovery;
+    private final int maxCommitAttempts;
+    private final boolean allowOutOfOrderCommits;
+    private final Integer timeoutSec;
+
+    private JdbcExactlyOnceOptions(
+            boolean discoverAndRollbackOnRecovery,
+            int maxCommitAttempts,
+            boolean allowOutOfOrderCommits,
+            Optional<Integer> timeoutSec) {
+        this.discoverAndRollbackOnRecovery = discoverAndRollbackOnRecovery;
+        this.maxCommitAttempts = maxCommitAttempts;
+        this.allowOutOfOrderCommits = allowOutOfOrderCommits;
+        this.timeoutSec = timeoutSec.orElse(null);
+        Preconditions.checkArgument(this.maxCommitAttempts > 0, "maxCommitAttempts should be > 0");
+    }
+
+    public static JdbcExactlyOnceOptions defaults() {
+        return builder().build();
+    }
+
+    public boolean isDiscoverAndRollbackOnRecovery() {
+        return discoverAndRollbackOnRecovery;
+    }
+
+    public boolean isAllowOutOfOrderCommits() {
+        return allowOutOfOrderCommits;
+    }
+
+    public int getMaxCommitAttempts() {
+        return maxCommitAttempts;
+    }
+
+    public Integer getTimeoutSec() {
+        return timeoutSec;
+    }
+
+    public static JDBCExactlyOnceOptionsBuilder builder() {
+        return new JDBCExactlyOnceOptionsBuilder();
+    }
+
+    /** JDBCExactlyOnceOptionsBuilder. */
+    public static class JDBCExactlyOnceOptionsBuilder {
+        private boolean recoveredAndRollback = DEFAULT_RECOVERED_AND_ROLLBACK;
+        private int maxCommitAttempts = DEFAULT_MAX_COMMIT_ATTEMPTS;
+        private boolean allowOutOfOrderCommits = DEFAULT_ALLOW_OUT_OF_ORDER_COMMITS;
+        private Optional<Integer> timeoutSec = Optional.empty();
+
+        public JDBCExactlyOnceOptionsBuilder withRecoveredAndRollback(
```

Comment:
Could you add Javadocs on the configuration mehtods about what are the configs used for?

Reply:
Sure, good point